pkg/router/gloo.go

```diff
@@ -0,0 +1,181 @@
+// GlooRouter is managing Istio virtual services
+type GlooRouter struct {
+	ugClient            gloov1.UpstreamGroupClient
+	logger              *zap.SugaredLogger
+	upstreamDiscoveryNs string
+}
+
+func NewGlooRouter(ctx context.Context, provider string, flaggerClient clientset.Interface, logger *zap.SugaredLogger, cfg *rest.Config) (*GlooRouter, error) {
+	// TODO if cfg is nil use memory client instead?
+	sharedCache := kube.NewKubeCache(ctx)
+	upstreamGroupClient, err := gloov1.NewUpstreamGroupClient(&factory.KubeResourceClientFactory{
+		Crd:             gloov1.UpstreamGroupCrd,
+		Cfg:             cfg,
+		SharedCache:     sharedCache,
+		SkipCrdCreation: true,
+	})
+	if err != nil {
+		// this should never happen.
+		return nil, fmt.Errorf("creating UpstreamGroup client %v", err)
+	}
+	if err := upstreamGroupClient.Register(); err != nil {
+		return nil, err
+	}
+	upstreamDiscoveryNs := ""
+	if strings.HasPrefix(provider, "gloo:") {
+		upstreamDiscoveryNs = strings.TrimPrefix(provider, "gloo:")
+	}
+
+	return NewGlooRouterWithClient(ctx, upstreamGroupClient, upstreamDiscoveryNs, logger), nil
+}
+
+func NewGlooRouterWithClient(ctx context.Context, routingRuleClient gloov1.UpstreamGroupClient, upstreamDiscoveryNs string, logger *zap.SugaredLogger) *GlooRouter {
+
+	if upstreamDiscoveryNs == "" {
+		upstreamDiscoveryNs = "gloo-system"
+	}
+	return &GlooRouter{ugClient: routingRuleClient, logger: logger, upstreamDiscoveryNs: upstreamDiscoveryNs}
+}
+
+// Reconcile creates or updates the Istio virtual service
+func (gr *GlooRouter) Reconcile(canary *flaggerv1.Canary) error {
+	return gr.SetRoutes(canary, 100, 0)
+}
+
+// GetRoutes returns the destinations weight for primary and canary
+func (gr *GlooRouter) GetRoutes(canary *flaggerv1.Canary) (
+	primaryWeight int,
+	canaryWeight int,
+	err error,
+) {
+	targetName := canary.Spec.TargetRef.Name
+	var ug *gloov1.UpstreamGroup
+	ug, err = gr.ugClient.Read(canary.Namespace, targetName, solokitclients.ReadOpts{})
+	if err != nil {
+		return
+	}
+
+	dests := ug.GetDestinations()
+	for _, dest := range dests {
+		if dest.GetDestination().GetUpstream().Name == upstreamName(canary.Namespace, fmt.Sprintf("%s-primary", targetName), canary.Spec.Service.Port) {
+			primaryWeight = int(dest.Weight)
+		}
+		if dest.GetDestination().GetUpstream().Name == upstreamName(canary.Namespace, fmt.Sprintf("%s-canary", targetName), canary.Spec.Service.Port) {
+			canaryWeight = int(dest.Weight)
+		}
+	}
+
+	if primaryWeight == 0 && canaryWeight == 0 {
+		err = fmt.Errorf("RoutingRule %s.%s does not contain routes for %s-primary and %s-canary",
+			targetName, canary.Namespace, targetName, targetName)
+	}
+
+	return
+}
+
+// SetRoutes updates the destinations weight for primary and canary
+func (gr *GlooRouter) SetRoutes(
+	canary *flaggerv1.Canary,
+	primaryWeight int,
+	canaryWeight int,
+) error {
+	targetName := canary.Spec.TargetRef.Name
+
+	destinations := []*gloov1.WeightedDestination{}
+	if primaryWeight != 0 {
```

Comment:
Why not create the route with zero weight? Doesn't Gloo support that?

Reply:
Just style. gloo supports 0 weight. I can change it if you prefer

Reply:
I think it would be better to have the same style as istio, appmesh and nginx, all have the zero weight route.